pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,197 @@
+		return errors.New("cluster controlPlaneConfiguration.Endpoint.Host is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Domain) <= 0 {
+		return errors.New("CloudStackDatacenterConfig domain is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Zone.Value) <= 0 {
+		return errors.New("CloudStackDatacenterConfig zone is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Network.Value) <= 0 {
+		return errors.New("CloudStackDatacenterConfig network is not set or is empty")
+	}
+	if clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef == nil {
+		return errors.New("must specify machineGroupRef for control plane")
+	}
+	controlPlaneMachineConfig := clusterSpec.controlPlaneMachineConfig()
+
+	if clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return errors.New("must specify machineGroupRef for worker nodes")
+	}
+
+	workerNodeGroupMachineConfig, ok := v.machineConfigs[clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name]
+	if !ok {
+		return fmt.Errorf("cannot find CloudStackMachineConfig %v for worker nodes", clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name)
+	}
+
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef == nil {
+			return errors.New("must specify machineGroupRef for etcd machines")
+		}
+		etcdMachineConfig = clusterSpec.etcdMachineConfig()
+		if len(etcdMachineConfig.Spec.Users) <= 0 {
+			etcdMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+		}
+		if len(etcdMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
+			etcdMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+		}
+	}
+
+	if len(controlPlaneMachineConfig.Spec.Users) <= 0 {
+		controlPlaneMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+	}
+	if len(workerNodeGroupMachineConfig.Spec.Users) <= 0 {
+		workerNodeGroupMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+	}
+	if len(controlPlaneMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
+		controlPlaneMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+	}
+	if len(workerNodeGroupMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
+		workerNodeGroupMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+	}
+
+	err := v.validateControlPlaneHost(&clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host)
+	if err != nil {
+		return err
+	}
+
+	for _, machineConfig := range v.machineConfigs {
+		if machineConfig.Namespace != clusterSpec.Namespace {
+			return errors.New("CloudStackMachineConfig and Cluster objects must have the same namespace specified")
+		}
+	}
+	if clusterSpec.datacenterConfig.Namespace != clusterSpec.Namespace {
+		return errors.New("CloudStackDatacenterConfig and Cluster objects must have the same namespace specified")
```

Comment:
These errors could be more informative. Tell the user what you received so they know, explicitly, what the program doesn't like.

```go
fmt.Errorf("CloudStackDatacenterConfig and Cluster objects must have the same namespace: CloudstackDatacenterConfig namespace=%s; Cluster namespace=%s", clusterSpec.datacenterConfig.Namespace, clusterSpec.Namespace)
```